back/engines/commercial/multi_tenancy/lib/tasks/core/migrate_votes_to_reactions.rake

```diff
@@ -0,0 +1,138 @@
+# frozen_string_literal: true
+
+# NOTE: Only to be used for release that migrates Votes -> Reactions
+namespace :fix_existing_tenants do
+  desc 'Transform all core data relating to votes into reactions'
+  task migrate_votes_core: [:environment] do |_t, _args|
+    # Order by active first - to reduce downtime
+    tenants = Tenant.creation_finalized.with_lifecycle('active') + Tenant.creation_finalized.with_lifecycle('demo')
+    tenants.each do |tenant|
+      Apartment::Tenant.switch(tenant.schema_name) do
+        VotesToReactionMigrator.new(tenant).run_core
+      end
+    end
+  end
+
+  desc 'Transform all activity data relating to votes into reactions'
+  task migrate_votes_activities: [:environment] do |_t, _args|
+    Tenant.creation_finalized.each do |tenant|
+      Apartment::Tenant.switch(tenant.schema_name) do
+        VotesToReactionMigrator.new(tenant).run_activities
+      end
+    end
+  end
+end
+
+class VotesToReactionMigrator
+  def initialize(tenant)
+    @tenant = tenant
+  end
+
+  def run_core
+    Rails.logger.info "Processing core data for tenant #{@tenant.host}"
+    update_downvoting_feature_flag
+    update_initiatives_voting_threshold
+    update_smart_groups
+    update_permissions
+    update_notifications
+    update_email_campaigns
+  end
+
+  def run_activities
+    Rails.logger.info "Processing activities for tenant #{@tenant.host}"
+    update_activities
+  end
+
+  private
+
+  def update_downvoting_feature_flag
+    Rails.logger.info "SETTING: Disliking feature flag for #{@tenant.host}"
+    settings = AppConfiguration.instance.settings
+    settings['disable_disliking'] = settings['disable_downvoting']
+    AppConfiguration.instance.update! settings: settings
+  end
+
+  def update_initiatives_voting_threshold
+    Rails.logger.info "SETTING: initiatives->reacting_threshold for #{@tenant.host}"
+    settings = AppConfiguration.instance.settings
+    settings['initiatives']['reacting_threshold'] = settings['initiatives']['voting_threshold']
+    AppConfiguration.instance.update! settings: settings
+    Rails.logger.info "SAVED: initiatives->reacting_threshold for #{@tenant.host}"
+  end
+
+  def update_smart_groups
+    Rails.logger.info "UPDATING: Smart groups for #{@tenant.host}"
+    count = 0
+    Group.where(membership_type: 'rules').each do |group|
+      next unless group.rules.to_s.include? 'voted'
+
+      group.rules.each do |rule|
+        rule['predicate'].sub!('voted', 'reacted')
+      end
+      if group.save
+        count += 1
+      else
+        Rails.logger.error "SMART_GROUP_ERROR: #{group.errors.errors}"
```

Comment:
I think it should be fine on production because most logs are disabled. But if for example the SQL queries would be logged, then the error messages could be hard to find back. In other rake tasks we keep a list of errors and print all the errors at the end.

Reply:
True. I'll implement that here.